packages/commonwealth/client/scripts/controllers/server/contracts.ts

```diff
@@ -15,12 +21,160 @@ class ContractsController {
   public get initialized() {
     return this._initialized;
   }
+  public getByNickname(nickname: string) {
+    return this._store.getContractByNickname(nickname);
+  }
   public getByIdentifier(id) {
     return this._store.getById(id);
   }
+  public getByAddress(address: string) {
+    return this._store.getContractByAddress(address);
+  }
   public getByType(type: string) {
     return this._store.getContractByType(type);
   }
+  public getContractFactories() {
+    return this._store.getContractFactories();
+  }
+  public getCommunityContracts() {
+    return this._store.getCommunityContracts();
+  }
+
+  public async addContractAbi(
+    contract: Contract,
+    abi: Array<Record<string, unknown>>
+  ) {
+    const response: TypedResponse<CreateContractAbiResp> = await $.post(
+      `${app.serverUrl()}/createContractAbi`,
+      {
+        jwt: app.user.jwt,
+        contractId: contract.id,
+        abi,
+      }
+    );
+    const resultContract: ContractAttributes = response['result']['contract'];
+    const resultAbi: ContractAbiAttributes = response['result']['contractAbi'];
+    this.update(resultAbi.abi, resultContract);
+    return response;
+  }
+
+  public async checkFetchEtherscanForAbi(address: string) {
+    try {
+      const response: Response = await $.post(
+        `${app.serverUrl()}/etherscanAPI/fetchEtherscanContract`,
+        {
+          address,
+        }
+      );
+      console.log(response);
+      const resultContract: ContractAttributes = response['result']['contract'];
+      const resultAbi: ContractAbiAttributes =
+        response['result']['contractAbi'];
+      this.update(resultAbi.abi, resultContract);
+    } catch (err) {
+      console.log('Failed to fetch abi from etherscan', err);
+      throw new Error(err);
+    }
+  }
+
+  public async update(
+    contractAbi: Array<Record<string, unknown>>,
+    contractAttributes: ContractAttributes
+  ) {
+    // Update contract in store
+    if (this._store.getById(contractAttributes.id)) {
+      this._store.remove(this._store.getById(contractAttributes.id));
+    }
+    const {
+      id,
+      address,
+      chain_node_id,
+      type,
+      createdAt,
+      updatedAt,
+      decimals,
+      token_name,
+      symbol,
+      is_factory,
+      nickname,
+    } = contractAttributes;
+    this._store.add(
+      new Contract(
```

Comment:
Can we use named arguments here, inside an object? Rather than having to list them in order?